web/modules/custom/dpl_webmaster/src/Form/InstallOrUpdateModule.php

```diff
@@ -0,0 +1,283 @@
+      }
+    }
+    elseif (!empty($all_files['project_upload']) && $this->moduleHandler->moduleExists('file')) {
+      $validators = ['FileExtension' => ['extensions' => $this->archiverManager->getExtensions()]];
+      /** @var \Drupal\file\FileInterface|null $finfo */
+      $finfo = file_save_upload('project_upload', $validators, FALSE, 0, FileExists::Replace);
+      if (!$finfo) {
+        // Failed to upload the file. file_save_upload() calls
+        // \Drupal\Core\Messenger\MessengerInterface::addError() on failure.
+        return;
+      }
+      /** @var string $local_cache */
+      $local_cache = $finfo->getFileUri();
+    }
+
+    $directory = _update_manager_extract_directory();
+    try {
+      $archive = update_manager_archive_extract($local_cache, $directory);
+    }
+    catch (\Exception $e) {
+      $this->messenger()->addError($e->getMessage());
+      return;
+    }
+
+    $files = $archive->listContents();
+    if (!$files) {
+      $this->messenger()->addError($this->t('Provided archive contains no files.'));
+      return;
+    }
+
+    // Unfortunately, we can only use the directory name to determine the
+    // project name. Some archivers list the first file as the directory (i.e.,
+    // MODULE/) and others list an actual file (i.e., MODULE/README.TXT).
+    $project = strtok($files[0], '/\\');
+
+    $archive_errors = $this->moduleHandler->invokeAll('verify_update_archive', [$project, $local_cache, $directory]);
+    if (!empty($archive_errors)) {
+      $this->messenger()->addError(array_shift($archive_errors));
+      // @todo Fix me in D8: We need a way to set multiple errors on the same
+      //   form element and have all of them appear!
+      if (!empty($archive_errors)) {
+        foreach ($archive_errors as $error) {
+          $this->messenger()->addError($error);
+        }
+      }
+      return;
+    }
+
+    // Make sure the Updater registry is loaded.
+    drupal_get_updaters();
+
+    $project_location = $directory . '/' . $project;
+    try {
+      $updater = Updater::factory($project_location, $this->root);
+    }
+    catch (\Exception $e) {
+      $this->messenger()->addError($e->getMessage());
+      return;
+    }
+
+    try {
+      $project_title = Updater::getProjectTitle($project_location);
+    }
+    catch (\Exception $e) {
+      $this->messenger()->addError($e->getMessage());
+      return;
+    }
+
+    if (!$project_title) {
+      $this->messenger()->addError($this->t('Unable to determine %project name.', ['%project' => $project]));
+    }
+
+    if (!$updater instanceof Module) {
+      $this->messenger()->addError($this->t('%project is not a module.', ['%project' => $project]));
+      return;
+    }
+
+    if ($updater->isInstalled()) {
+      // Tell UpdateReady form which projects to update.
+      $this->session->set('update_manager_update_projects', [$project => $project]);
+      $form_state->setRedirect('update.confirmation_page');
+      return;
+    }
```

Comment:
Consider adding a code comments adding "markers" to  explain where the deviation from `UpdateManagerInstall` starts and ends.

This approach makes it hard to determine what has been changed compared to the original - and thus also hard to review.